src/sinks/axiom.rs

```diff
@@ -0,0 +1,397 @@
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    config::{
+        AcknowledgementsConfig, GenerateConfig, Input, SinkConfig, SinkContext, SinkDescription,
+    },
+    sinks::{
+        elasticsearch::{ElasticsearchAuth, ElasticsearchConfig},
+        util::{http::RequestConfig, Compression},
+        Healthcheck, VectorSink,
+    },
+    tls::TlsConfig,
+};
+
+static CLOUD_URL: &str = "https://cloud.axiom.co";
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub(self) struct AxiomConfig {
+    url: Option<String>,
+    org_id: Option<String>,
+    token: String,
+    dataset: String,
+
+    #[serde(default)]
+    compression: Compression,
+    tls: Option<TlsConfig>,
+    #[serde(
+        default,
+        deserialize_with = "crate::serde::bool_or_struct",
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    acknowledgements: AcknowledgementsConfig,
+}
+
+inventory::submit! {
+    SinkDescription::new::<AxiomConfig>("axiom")
+}
+
+impl GenerateConfig for AxiomConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(
+            r#"token = "${AXIOM_TOKEN}"
+            dataset = "my-dataset"
+            url = "${AXIOM_URL}"
+            org_id = "${AXIOM_ORG_ID}""#,
+        )
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "axiom")]
+impl SinkConfig for AxiomConfig {
+    async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+        let mut headers = IndexMap::with_capacity(1);
+        headers.insert(
+            "X-Axiom-Org-Id".to_string(),
+            self.org_id.clone().unwrap_or_default(),
+        );
+        let mut query = HashMap::with_capacity(1);
+        query.insert("timestamp-field".to_string(), "timestamp".to_string());
+        let elasticsearch_config = ElasticsearchConfig {
+            endpoint: self.build_endpoint(),
+            compression: self.compression,
+            auth: Some(ElasticsearchAuth::Basic {
+                user: "axiom".to_string(),
+                password: self.token.clone(),
+            }),
+            query: Some(query),
+            tls: self.tls.clone(),
+            request: RequestConfig {
+                headers,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+        let config = elasticsearch_config.build(cx).await?;
+        Ok(config)
+    }
+
+    fn input(&self) -> Input {
+        Input::log()
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "axiom"
+    }
+
+    fn acknowledgements(&self) -> Option<&AcknowledgementsConfig> {
+        Some(&self.acknowledgements)
+    }
+}
+
+impl AxiomConfig {
+    fn build_endpoint(&self) -> String {
+        let url = if let Some(url) = self.url.as_ref() {
+            url.clone()
+        } else {
+            CLOUD_URL.to_string()
+        };
+
+        format!("{}/api/v1/datasets/{}/elastic", url, self.dataset)
```

Comment:
I guess it'd be useful for some people, yeah. What do I need to do to enable that?

Reply:
We could make the field a `template` instead, `elasticsearch` has a few examples: https://github.com/vectordotdev/vector/blob/master/src/sinks/elasticsearch/config.rs

AWS S3, also has `key_prefix` as a `template`: https://github.com/vectordotdev/vector/blob/master/src/sinks/aws_s3/config.rs

Reply:
I started implementing this, but because the dataset is part of the bulk endpoint, we'd have to make the `endpoint` in the Elasticsearch sink a template which doesn't work well with the current architecture afaict. 

With all that I don't think it's worth it at this time, people could always use multiple filter transforms to send data to multiple datasets.

Reply:
Sounds great, thanks!

Reply:
Ah - good catch, didn't realize that on first glance. ☝️ sounds good to me!